### PR TITLE
enable external oauth clients

### PR DESCRIPTION
**Pull request by minrk**
Built on #1577

TODO:

- [x] improve tracking of token activity (this gets increasingly important as we expand scope of how tokens can be used)
- [x] alembic upgrade
- [x] support defining oauth clients in config
- [x] allow oauth to come from external (specified) hosts

Once complete, this will allow JupyterHub to serve as a 'whoami' OAuth provider. OAuth tokens do not grant access to perform actions on behalf of users.

Logical extensions not in this PR:

- implement oauth scopes (requires scope-approval page), allowing OAuth tokens to take action. This will probably mean merging OAuth tokens with the existing APIToken. They are only separate now because OAuth tokens do not grant API access.
- allow registering new OAuth clients via UI, not just config.

### CODE REVIEW

**Comment by willingc** [comment]
@minrk when this pr lands, will nbgrader then work with master and 0.8.1 (assuming backward compatibility_?

File: jupyterhub/apihandlers/auth.py [hunk 1]
```diff
@@ -49,7 +53,7 @@ def post(self):
                 raise web.HTTPError(403)
         else:
             data = self.get_json_body()
-            # admin users can request 
+            # admin users can request tokens for other usrs
```
Comment (willingc) [inline jupyterhub/apihandlers/auth.py]: typo users

**Comment by minrk** [comment]
This shouldn't affect backward compatibility, so nothing should change for nbgrader between 0.8.x and this. It's purely a new feature for allowing oauth clients to have redirect URIs other than `/services/:name/oauth_callback`.

**Comment by yuvipanda** [comment]
w00t, this is awesome!

**Comment by yuvipanda** [comment]
This is awesome! This should also allow us do authenticated binderhubs I think?

**Comment by minrk** [comment]
> This should also allow us do authenticated binderhubs I think?

It's a big part of it, yes. However, binder deployment would also need updates to make the nullauthenticator bit optional.

File: jupyterhub/alembic/versions/56cc5a70207e_token_tracking.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+branch_labels = None
+depends_on = None
+
+from alembic import op
+import sqlalchemy as sa
+
+
+def upgrade():
+    tables = op.get_bind().engine.table_names()
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (willingc) [inline jupyterhub/alembic/versions/56cc5a70207e_token_tracking.py]: Under what circumstances would this be adjusted? Different data fields stored for tokens?

File: jupyterhub/app.py [hunk 3]
```diff
@@ -1145,7 +1145,8 @@ def init_services(self):
             if service.managed:
                 if not service.api_token:
                     # generate new token
-                    service.api_token = service.orm.new_api_token()
+                    # TODO: revoke old tokens?
```
Comment (willingc) [inline jupyterhub/app.py]: Are there use cases where a managed service would need to keep an old api_token?

File: jupyterhub/apihandlers/auth.py [hunk 4]
```diff
@@ -31,33 +33,43 @@ def get(self, token):
...
         else:
             data = self.get_json_body()
-            # admin users can request 
-            if data and data.get('username') != user.name:
-                if user.admin:
-                    user = self.find_user(data['username'])
-                    if user is None:
-                        raise web.HTTPError(400, "No such user '%s'" % data['username'])
-                else:
+            # admin users can request tokens for other users
```
Comment (willingc) [inline jupyterhub/apihandlers/auth.py]: can an admin user request a token for another admin user? I assume that is okay.

File: jupyterhub/app.py [hunk 5]
```diff
@@ -1266,6 +1269,26 @@ def init_oauth(self):
             login_url=url_path_join(base_url, 'login')
         )
 
+    def cleanup_oauth_clients(self):
+        """Cleanup any OAuth clients that shouldn't be in the database.
+
+        This should mainly be services that have been removed from configuration or renamed.
```
Comment (willingc) [inline jupyterhub/app.py]: +1

File: jupyterhub/orm.py [hunk 6]
```diff
@@ -319,6 +321,11 @@ def service_id(cls):
     hashed = Column(Unicode(255), unique=True)
     prefix = Column(Unicode(16), index=True)
 
+    # token metadata for bookkeeping
```
Comment (willingc) [inline jupyterhub/orm.py]: +1

**Comment by yuvipanda** [comment]
Not necessarily in this PR, but would be also great to have an example here!

File: jupyterhub/app.py (hunk 3 above)
Comment (minrk) [reply jupyterhub/app.py]: Not one generated here. I think the main bit missing is that we don't track whether this happened, so we don't have enough info to clean it up afterward.

File: jupyterhub/apihandlers/auth.py (hunk 4 above)
Comment (minrk) [reply jupyterhub/apihandlers/auth.py]: Yes, any other user.

File: jupyterhub/alembic/versions/56cc5a70207e_token_tracking.py (hunk 2 above)
Comment (minrk) [reply jupyterhub/alembic/versions/56cc5a70207e_token_tracking.py]: This is an auto-generated comment from `alembic revision --autogenerate`. I should have removed it after making sure it was correct.

**Comment by minrk** [comment]
external oauth example added, review addressed. Should be good to go.

File: examples/external-oauth/README.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
+# Using JupyterHub as an OAuth provider
+
+JupyterHub 0.9 introduces
+Uses `jupyterhub.services.HubAuthenticated` to authenticate requests with the Hub.
```
Comment (willingc) [inline examples/external-oauth/README.md]: ~~Uses~~ using

File: examples/external-oauth/README.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+1. generate an API token:
+
+        export JUPYTERHUB_API_TOKEN=`openssl rand -hex 32`
+
+2. launch the whoami service:
+
+        bash launch-service.sh &
+
+3. Launch JupyterHub:
+
```
Comment (willingc) [inline examples/external-oauth/README.md]: command??

File: examples/external-oauth/README.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
+# Using JupyterHub as an OAuth provider
+
+JupyterHub 0.9 introduces
+Uses `jupyterhub.services.HubAuthenticated` to authenticate requests with the Hub.
+
+There is an implementation each of cookie-based `HubAuthenticated` and OAuth-based `HubOAuthenticated`.
+
+## Run
```
Comment (willingc) [inline examples/external-oauth/README.md]: Running JupyterHub as an OAuth provider

**Pull request merged**